### PR TITLE
Certificate manager verify callback

### PR DESCRIPTION
**Pull request by embhorn**
Adds the verifyCallback function in the Certificate Manager. If present, it will be called for each cert verified using `wolfSSL_CertManagerVerify`. If there is a verification error, the verify callback can be used to over-ride the error.

Test with https://github.com/embhorn/wolfssl-examples/tree/cert_vfy_CB

### CODE REVIEW

File: src/ssl.c [hunk 1]
```diff
@@ -5867,6 +5867,136 @@ int wolfSSL_CertManagerDisableCRL(WOLFSSL_CERT_MANAGER* cm)
...
+            XMEMCPY(domain, cert->subjectCN, subjectCNLen);
+            domain[subjectCNLen] = '\0';
+        }
+    }
+
+    store->error = ret;
+    store->error_depth = 1;
+    store->discardSessionCerts = 0;
+    store->domain = domain;
+    store->totalCerts = 1;
```
Comment (dgarske) [inline src/ssl.c]: Did you consider having a user supplied `void*` context? Perhaps even just setting `store->userCtx = cm`, so the verify callback can access the certificate manager?

File: src/ssl.c [hunk 2]
```diff
@@ -5867,6 +5867,136 @@ int wolfSSL_CertManagerDisableCRL(WOLFSSL_CERT_MANAGER* cm)
...
+void wolfSSL_CertManagerSetVerify(WOLFSSL_CERT_MANAGER* cm, VerifyCallback vc)
+{
+    WOLFSSL_ENTER("wolfSSL_CertManagerSetVerify");
+    if (cm == NULL)
+        return;
+
+    cm->verifyCallback = vc;
+}
+
+static int CertManager_DoVerifyCallback(WOLFSSL_CERT_MANAGER* cm, int ret,
```
Comment (dgarske) [inline src/ssl.c]: The code here is very similar to internal.c `DoVerifyCallback`. Did you consider combining any of that logic into a function both can use? I did a review and its possible, but the added logic for handling cm / ssl may not reduce the code size. Just something to consider.

File: src/ssl.c [hunk 3]
```diff
@@ -5910,6 +6040,11 @@ int wolfSSL_CertManagerVerifyBuffer(WOLFSSL_CERT_MANAGER* cm, const byte* buff,
     if (ret == 0)
         ret = ParseCertRelative(cert, CERT_TYPE, 1, cm);
 
+    /* if verify callback has been set */
```
Comment (dgarske) [inline src/ssl.c]: The calls to `DoVerifyCallback` perform the callback after the OCSP/CRL checking. Perhaps this callback should be below those?

Comment (embhorn) [reply src/ssl.c]: Agreed.

File: src/ssl.c (hunk 2 above)
Comment (embhorn) [reply src/ssl.c]: Yes, but as you mentioned, the extra handling involved with the SSL structure makes would make the code difficult to follow with little benefit.

File: src/ssl.c (hunk 1 above)
Comment (embhorn) [reply src/ssl.c]: I like this approach. The CM pointer could be useful for the callback.

File: src/ssl.c [hunk 4]
```diff
@@ -5867,6 +5867,136 @@ int wolfSSL_CertManagerDisableCRL(WOLFSSL_CERT_MANAGER* cm)
...
     return WOLFSSL_SUCCESS;
 }
+
+void wolfSSL_CertManagerSetVerify(WOLFSSL_CERT_MANAGER* cm, VerifyCallback vc)
+{
+    WOLFSSL_ENTER("wolfSSL_CertManagerSetVerify");
+    if (cm == NULL)
+        return;
+
+    cm->verifyCallback = vc;
```
[the remark this answers is not in the document]
Comment (embhorn) [reply src/ssl.c]: Used your suggestion below.

**Comment by embhorn** [comment]
Great feedback, thank you. I added a test case that overrides an expired certificate using the verify CB.

File: wolfssl/internal.h [hunk 5]
```diff
@@ -1952,6 +1952,9 @@ struct WOLFSSL_CERT_MANAGER {
 #endif
     char*           ocspOverrideURL;     /* use this responder */
     void*           ocspIOCtx;           /* I/O callback CTX */
+#ifndef NO_WOLFSSL_CM_VERIFY
+    VerifyCallback  verifyCallback;      /* Verify callback */
```
Comment (toddouska) [inline wolfssl/internal.h]: This seems correct but then isn't the VerifyCallback in WOLFSSL_CTX redundant?  Or does one have precedence if they're both on?

Comment (embhorn) [reply wolfssl/internal.h]: It could be a different callback function altogether. It allows greater flexibility when over riding cert manager errors.

File: src/ssl.c [hunk 6]
```diff
@@ -6153,6 +6153,139 @@ int wolfSSL_CertManagerDisableCRL(WOLFSSL_CERT_MANAGER* cm)
...
+    XMEMSET(store->param, 0, sizeof(WOLFSSL_X509_VERIFY_PARAM));
+#endif /* defined(OPENSSL_EXTRA) || defined(HAVE_WEBSERVER)*/
+
+#if defined(OPENSSL_EXTRA) || defined(OPENSSL_EXTRA_X509_SMALL)
+    InitX509(x509, 0, cm->heap);
+    if (CopyDecodedToX509(x509, cert) == 0) {
+        store->current_cert = x509;
+    }
+    else {
+        FreeX509(x509);
```
Comment (toddouska) [inline src/ssl.c]: Isn't this already called on 6267 below?

File: src/ssl.c [hunk 7]
```diff
@@ -6201,6 +6334,13 @@ int wolfSSL_CertManagerVerifyBuffer(WOLFSSL_CERT_MANAGER* cm, const byte* buff,
         ret = CheckCertCRL(cm->crl, cert);
 #endif
 
+#ifndef NO_WOLFSSL_CM_VERIFY
+    /* if verify callback has been set */
+    if (cm->verifyCallback) {
+        ret = CertManager_DoVerifyCallback(cm, ret, cert);
```
Comment (toddouska) [inline src/ssl.c]: This is called for each CM verify.  Your commit comment says for each cert loaded.  What about wolfSSL_CertManagerLoadCA()?  Is that not covered by design?

File: src/ssl.c (hunk 6 above)
Comment (embhorn) [reply src/ssl.c]: Fixed.

File: src/ssl.c (hunk 7 above)
Comment (embhorn) [reply src/ssl.c]: I can revise the wording in the commit. The API currently relies on the application to call `wolfSSL_CertManagerVerify` for each loaded certificate.

File: src/ssl.c [hunk 8]
```diff
@@ -6685,6 +6889,12 @@ int wolfSSL_CTX_load_verify_locations_ex(WOLFSSL_CTX* ctx, const char* file,
     }
 
     verify = GET_VERIFY_SETTING_CTX(ctx);
+#ifndef NO_WOLFSSL_CM_VERIFY
+    if (flags & WOLFSSL_LOAD_FLAG_USE_CB) {
+        verify = VERIFY_USE_CB;
```
Comment (dgarske) [inline src/ssl.c]: This will override the above "verify" settings, which is not okay. Why is this required at all? Isn't it sufficient to just check "cm-verifyCallback" is set?

File: src/ssl.c [hunk 9]
```diff
@@ -6659,6 +6797,65 @@ int ProcessFile(WOLFSSL_CTX* ctx, const char* fname, int format, int type,
...
+            InitDecodedCert(cert, der->buffer, der->length, ctx->cm->heap);
+    #else
+            tmp_ret = NOT_COMPILED_IN;
+    #endif
+        }
+        else {
+            InitDecodedCert(cert, (byte*)myBuffer, (word32)sz, ctx->cm->heap);
+        }
+        if (tmp_ret == 0)
+            tmp_ret = ParseCertRelative(cert, type, verify, ctx->cm);
```
Comment (toddouska) [inline src/ssl.c]: Where is this being checked?

File: src/ssl.c [hunk 10]
```diff
@@ -6659,6 +6797,65 @@ int ProcessFile(WOLFSSL_CTX* ctx, const char* fname, int format, int type,
                                 userChain, verify);
     }
 
+#ifndef NO_WOLFSSL_CM_VERIFY
+    /* Call to over-ride status */
+    if ((ctx != NULL) && (ctx->cm != NULL) &&
+        (ctx->cm->verifyCallback != NULL)) {
+        int tmp_ret = 0;
```
Comment (toddouska) [inline src/ssl.c]: There are keys, encrypted keys, crls, and user chain certs at the least here that I don't see you removing.  That is, trying to parse a private key and calling cert verify override on it should be breaking lots of cases.  Something is wrong here.

File: src/ssl.c [hunk 11]
```diff
@@ -6647,9 +6802,18 @@ int ProcessFile(WOLFSSL_CTX* ctx, const char* fname, int format, int type,
...
         if ((type == CA_TYPE || type == TRUSTED_PEER_TYPE)
-                                              && format == WOLFSSL_FILETYPE_PEM)
+                                          && format == WOLFSSL_FILETYPE_PEM) {
             ret = ProcessChainBuffer(ctx, myBuffer, sz, format, type, ssl,
                                      verify);
+#ifndef NO_WOLFSSL_CM_VERIFY
+            /* Call to over-ride status */
+            if ((ctx != NULL) && (ctx->cm != NULL) &&
+                (ctx->cm->verifyCallback != NULL)) {
+                ret = wolfSSL_CertManagerVerifyBuffer_ex(ctx->cm, myBuffer, sz,
```
Comment (dgarske) [inline src/ssl.c]: This will also use the CM's verify callback for any CA or Trusted peer type certs when `ProcessFile` is called. I don't see this described in the PR comments or any test cases. Does this addition really need to be here?

File: tests/api.c [hunk 12]
```diff
@@ -1143,6 +1143,43 @@ static void test_wolfSSL_CertManagerGetCerts(void)
...
+    AssertNotNull(cm);
+
+    /* Use the test CB that always accepts certs */
+    myVerifyFail = 0;
+    wolfSSL_CertManagerSetVerify(cm, myVerify);
+
+    ret = wolfSSL_CertManagerLoadCA(cm, ca_cert, NULL);
+    AssertIntEQ(ret, WOLFSSL_SUCCESS);
+
+    ret = wolfSSL_CertManagerVerify(cm, ca_expired_cert, WOLFSSL_FILETYPE_PEM);
```
Comment (dgarske) [inline tests/api.c]: The `wolfSSL_CertManagerVerify` API should be used for non CA certificates (type = `CERT_TYPE`) (see ssl.c:6335). The test case should not be using a "CA".

File: src/ssl.c [hunk 13]
```diff
@@ -6153,9 +6154,146 @@ int wolfSSL_CertManagerDisableCRL(WOLFSSL_CERT_MANAGER* cm)
...
+void wolfSSL_CertManagerSetVerify(WOLFSSL_CERT_MANAGER* cm, VerifyCallback vc)
+{
+    WOLFSSL_ENTER("wolfSSL_CertManagerSetVerify");
+    if (cm == NULL)
+        return;
+
+    cm->verifyCallback = vc;
+}
+
+static int CertManager_DoVerifyCallback(WOLFSSL_CERT_MANAGER* cm, int ret,
```
Comment (dgarske) [inline src/ssl.c]: In the ideal world this function should be combined with internal.c `DoVerifyCallback` to eliminate duplicate code. If you rebase with latest master the `DoVerifyCallback` function has been improved to not do the SSL alert, you may find its possible to combine these now.
